test/integration/odcds_integration_test.cc

```diff
@@ -0,0 +1,415 @@
+                      api_type: DELTA_GRPC
+                      transport_api_version: V3
+                      grpc_services:
+                        envoy_grpc:
+                          cluster_name: xds_cluster
+              routes:
+              - match: {{ prefix: "/" }}
+                route:
+                  cluster_header: "Pick-This-Cluster"
+)EOF",
+                                                  Platform::null_device_path));
+}
+
+class OdCdsIntegrationTestBase : public HttpIntegrationTest,
+                                 public Grpc::GrpcClientIntegrationParamTest {
+public:
+  OdCdsIntegrationTestBase(const std::string& config)
+      : HttpIntegrationTest(Http::CodecClient::Type::HTTP2, ipVersion(), config) {
+    use_lds_ = false;
+  }
+
+  void TearDown() override {
+    if (doCleanUpXdsConnection_) {
+      cleanUpXdsConnection();
+    }
+  }
+
+  void initialize() override {
+    // Controls how many addFakeUpstream() will happen in
+    // BaseIntegrationTest::createUpstreams() (which is part of initialize()).
+    // Make sure this number matches the size of the 'clusters' repeated field in the bootstrap
+    // config that you use!
+    setUpstreamCount(1);                                  // The xDS cluster
+    setUpstreamProtocol(FakeHttpConnection::Type::HTTP2); // CDS uses gRPC uses HTTP2.
+
+    // BaseIntegrationTest::initialize() does many things:
+    // 1) It appends to fake_upstreams_ as many as you asked for via setUpstreamCount().
+    // 2) It updates your bootstrap config with the ports your fake upstreams are actually listening
+    //    on (since you're supposed to leave them as 0).
+    // 3) It creates and starts an IntegrationTestServer - the thing that wraps the almost-actual
+    //    Envoy used in the tests.
+    // 4) Bringing up the server usually entails waiting to ensure that any listeners specified in
+    //    the bootstrap config have come up, and registering them in a port map (see lookupPort()).
+    HttpIntegrationTest::initialize();
+
+    addFakeUpstream(FakeHttpConnection::Type::HTTP2);
+    new_cluster_ = ConfigHelper::buildStaticCluster(
+        "new_cluster", fake_upstreams_[1]->localAddress()->ip()->port(),
+        Network::Test::getLoopbackAddressString(ipVersion()));
+
+    test_server_->waitUntilListenersReady();
+    registerTestServerPorts({"http"});
+  }
+
+  FakeStreamPtr odcds_stream_;
+  envoy::config::cluster::v3::Cluster new_cluster_;
+  bool doCleanUpXdsConnection_ = true;
+};
+
+class OdCdsIntegrationTest : public OdCdsIntegrationTestBase {
+public:
+  OdCdsIntegrationTest() : OdCdsIntegrationTestBase(config()) {}
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersionsClientType, OdCdsIntegrationTest,
+                         GRPC_CLIENT_INTEGRATION_PARAMS);
+
+// tests a scenario when:
+//  - making a request to an unknown cluster
+//  - odcds initiates a connection with a request for the cluster
+//  - a response contains the cluster
```

Comment:
We should also add a timeout case.

Reply:
Also a case for when the cluster is not available at the server.

Reply:
Added the test cases.